[openmp][AIX]Initial changes for porting to AIX

xingxue-ibm opened the pull request:
This PR contains initial changes for building and testing libomp on AIX. More changes will follow.
- `KMP_OS_AIX` is defined for the AIX platform
- `KMP_ARCH_PPC` is defined for 32-bit PPC
- `KMP_ARCH_PPC_XCOFF` and `KMP_ARCH_PPC64_XCOFF` are for 32- and 64-bit XCOFF object formats respectively
- Assembly file `z_AIX_asm.S` is used for AIX specific assembly code and will be added in a separate PR
- The target library is disabled because AIX does not have the device support
- OMPT is temporarily disabled

xingxue-ibm committed:
Initial changes for AIX.

xingxue-ibm committed:
git-clang-format changes.

kkwli commented on openmp/runtime/CMakeLists.txt:
```diff
@@ -89,7 +91,7 @@ if(LIBOMP_ARCH STREQUAL "aarch64")
 endif()
 
-libomp_check_variable(LIBOMP_ARCH 32e x86_64 32 i386 arm ppc64 ppc64le aarch64 aarch64_a64fx mic mips mips64 riscv64 loongarch64 ve s390x wasm32)
+libomp_check_variable(LIBOMP_ARCH 32e x86_64 32 i386 arm ppc64 ppc64le aarch64 aarch64_a64fx mic mips mips64 riscv64 loongarch64 ve s390x wasm32 ppc)
```
Nit: Group `ppc*` to together.

xingxue-ibm replied:
Changed as suggested, thanks!

kkwli commented on openmp/runtime/CMakeLists.txt:
```diff
@@ -30,7 +30,7 @@ if(${OPENMP_STANDALONE_BUILD})
   libomp_get_architecture(LIBOMP_DETECTED_ARCH)
   set(LIBOMP_ARCH ${LIBOMP_DETECTED_ARCH} CACHE STRING
-    "The architecture to build for (x86_64/i386/arm/ppc64/ppc64le/aarch64/mic/mips/mips64/riscv64/loongarch64/ve/s390x/wasm32).")
+    "The architecture to build for (x86_64/i386/arm/ppc64/ppc64le/aarch64/mic/mips/mips64/riscv64/loongarch64/ve/s390x/wasm32/ppc).")
```
Nit: Group `ppc*` together

xingxue-ibm replied:
Changed as suggested, thanks!

kkwli commented on openmp/cmake/OpenMPTesting.cmake:
```diff
@@ -55,6 +55,9 @@ if (${OPENMP_STANDALONE_BUILD})
   if (MSVC OR XCODE)
     set(DEFAULT_LIT_ARGS "${DEFAULT_LIT_ARGS} --no-progress-bar")
   endif()
+  if (${CMAKE_SYSTEM_NAME} MATCHES "AIX")
```
If we want to impose time limit for the lit tests, should the limit also apply to all platform?

xingxue-ibm replied:
The time limit depends on the platform so I'd refrain from setting it for other platforms.

kkwli commented on openmp/runtime/CMakeLists.txt:
```diff
@@ -203,6 +211,8 @@ elseif("${LIBOMP_ARCH}" STREQUAL "s390x") # S390x (Z) architecture
     set(S390X TRUE)
 elseif("${LIBOMP_ARCH}" STREQUAL "wasm32") # WebAssembly architecture
     set(WASM TRUE)
+elseif("${LIBOMP_ARCH}" STREQUAL "ppc") # PPC32 architecture
```
Nit: Move this closer to line 191

xingxue-ibm replied:
Changed as suggested, thanks!

xingxue-ibm committed:
Addressed comments.
- group ppc closer to ppc64

shiltian commented on openmp/runtime/CMakeLists.txt:
```diff
@@ -128,8 +130,13 @@ set(LIBOMP_ASMFLAGS "" CACHE STRING
-set(LIBOMP_LIBFLAGS "" CACHE STRING
-  "Appended user specified linked libs flags. (e.g., -lm)")
+if("${LIBOMP_ARCH}" STREQUAL "ppc")  # PPC (32-bit) needs libatomic for __atomic_load_8, etc.
+set(LIBOMP_LIBFLAGS "-latomic" CACHE STRING
```
indent

brad0 replied:
Add AIX OS check here.

shiltian commented on openmp/runtime/src/CMakeLists.txt:
```diff
@@ -108,7 +108,11 @@ else()
     libomp_append(LIBOMP_CXXFILES kmp_gsupport.cpp)
-    libomp_append(LIBOMP_GNUASMFILES z_Linux_asm.S) # Unix assembly file
+    if(${CMAKE_SYSTEM_NAME} MATCHES "AIX")
+      libomp_append(LIBOMP_GNUASMFILES z_AIX_asm.S) # AIX assembly file
```
It looks like `z_AIX_asm.S` is missing in this patch

brad0 replied:
@shiltian It says in the description that that portion would be added in a separate PR.

shiltian replied:
Oh, sorry for missing that part. If that is the case, that PR needs to be landed before this one, otherwise it will cause issues. However, I don't know how that patch would be tested w/o this one. That being said, I think it might make more sense to include it in this patch as well.

xingxue-ibm replied:
We plan to upstream changes in stages. This PR contains initial changes and lays the ground for subsequent AIX port PRs. The PR for `z_AIX_asm.S` will be posted shortly. Since the buildbot of AIX libomp is not set up yet and the code is guarded for AIX, this shouldn't cause problems. Our downstream build has LIT tests clean and we are still enabling affinity and some other things.

shiltian commented on openmp/CMakeLists.txt:
```diff
@@ -94,7 +94,8 @@ set(ENABLE_LIBOMPTARGET ON)
 # 32-bit systems are not supported either.
-if (APPLE OR WIN32 OR WASM OR NOT "cxx_std_17" IN_LIST CMAKE_CXX_COMPILE_FEATURES OR NOT CMAKE_SIZEOF_VOID_P EQUAL 8)
+if (APPLE OR WIN32 OR WASM OR NOT "cxx_std_17" IN_LIST CMAKE_CXX_COMPILE_FEATURES
+    OR NOT CMAKE_SIZEOF_VOID_P EQUAL 8 OR ${CMAKE_SYSTEM_NAME} MATCHES "AIX")
```
I see many places have `${CMAKE_SYSTEM_NAME} MATCHES "AIX"`. Can we just define a variable (e.g. `AIX`, like others) and use it everywhere else?

brad0 commented on openmp/runtime/src/z_Linux_util.cpp:
```diff
@@ -62,6 +64,18 @@
+  {                                                                            \
+    (tv)->tv_sec = (ts)->tv_sec;                                               \
+    (tv)->tv_usec = (ts)->tv_nsec / 1000;                                      \
+  }
```
TIMESPEC_TO_TIMEVAL is not used in this code and there is a copy of TIMEVAL_TO_TIMESPEC just a few lines below this.

xingxue-ibm replied:
Good catch, thanks!

xingxue-ibm committed:
Addressed comments.
- Add check for AIX for adding libatomic for 32-bit PPC
- Remove macro definitions TIMEVAL_TO_TIMESPEC and TIMESPEC_TO_TIMEVAL
  for AIX.

shiltian reviewed (approved):
The patch generally looks good to me. We can improve things as we go.

brad0 reviewed (approved):
LGTM as a start towards AIX support.

xingxue-ibm commented:
Thanks very much, @brad0, @shiltian, @kkwli for reviewing this PR!
